### DISCUSSION TITLE
Simple Window Switcher Thread

### DISCUSSION THREAD

**Discussion by valinet** [valinet/ExplorerPatcher#110] [Show and tell]
Hi

I am pleased to introduce a new, special feature in the latest release of ExplorerPatcher: Simple Window Switcher (sws), as it is called internally, or the "Windows 10 window switcher" aka the Alt+Tab window from Windows 10.

![image](https://user-images.githubusercontent.com/6503598/137497130-7afe73c6-e324-49fa-a03e-7c95abc350b8.png)

It is a beautiful, clean, written-from-scratch implementation of a window switcher replicating the one used in Windows 10. It offers a couple of configuration options, which I will let you discover on your own, and it does not get in the way. It is fast, and behaves in a natural way.

For the moment, the feature is disabled by default. To enable it, press `Win`+`X` > Properties > Window switcher > Enable Windows 10 window switcher.

![image](https://user-images.githubusercontent.com/6503598/137500712-41d018f5-2ebe-42c1-a1d1-cf89c59063c0.png)

For those interested, it is entirely written in C, with the source code, of course, fully available here: https://github.com/valinet/sws. It’s pulled in as a submodule by the main repo and statically compiled directly in the main library.

Known issues:
* ~~Probably the selection rectangle when using the light theme is not dark enough~~ Solved
* ~~Some apps hang the application when trying to close them; we need a better API to close windows:~~
  * `SendMessage(hWnd, WM_CLOSE, 0, 0)` does not work reach process at higher privilege levels
  * `EndTask` (currently used) works but it is prone to locking when the application displays some UI before closing, or in other situations
  * Solved. The better solution is to use `PostMessage(hWnd, WM_SYSCOMMAND, SC_CLOSE, 0)` which works on windows with that window style, which is a condition to be included in Alt+Tab anyway, so it works rather nicely for all windows shown and cross integrity level limits, apparently 
* ~~Windows in bands superior to (or other than) `ZBID_DESKTOP` always are displayed first in the tab order. That's because the legacy window enumeration APIs enumerate window in Z-order, not in activation order. Alt+Tab in Explorer uses some other undocumented API, We have to discover and use that as well. ~~Similarly for closing a window, probably.~~ A relevant thread here: https://stackoverflow.com/questions/69467597/win32-enumerate-real-windows-in-real-activation-order~~ Solved: The window switcher now tracks window activations internally.

Fixed:
* As opposed to the default Alt+Tab view, which crashes Explorer if a window closes randomly while the UI is displayed, this one just gracefully handles that situation.

What will probably be never implemented:
* Microsoft Edge tabs as windows, because I HATE that feature. `Alt`+`Tab` is for windows only.

Thanks

**Comment by MonkAlex** [comment]
Full screen youtube page in firefox (93.0) not enumerated.

And yeah, 
> Probably the selection rectangle when using the light theme is not dark enough

Its not nice visible in any theme.

**Reply by valinet** [reply]
First issue should be fixed.
Second one, I think in dark theme it being white looks fine. In light theme it’s an issue maybe indeed, looking to fix it up.

**Comment by Gaurav-Original-ClassicShellTester** [comment]
Well, the window switcher does look great and very authentic like the Windows 10 switcher and it can show a whopping 12 apps on one page at a time on my 15 inch 16:9 laptop screen if I want it which is just awesome (probably more on my 17-inch?, haven't tried it there as I have only Windows 10 there).

![image](https://user-images.githubusercontent.com/46422834/138551875-dcd97ff3-cea7-4f7c-9d4c-aee6348c1508.png)

And the stock Windows 11 Alt-Tab incorrectly shows a hidden window from one of Microsoft's OWN UWP apps - their Miracast Wireless Display Adapter app that I can't seem to hide in the stock switcher. But it hides perfectly in ExplorerPatcher's switcher.

And it includes the Desktop as a switchable item. Awesome work really since Microsoft has been fooling around with Alt-Tab in every release since Vista and even when they get it right, they ruin it again.

**Reply by valinet** [reply]
Thank you 🙏

**Comment by ShiroUwUs7** [comment]
Only thing I've noticed is some x64 apps don't render in the alt+tab window, not sure if its related or just an issue on my side. 
But other then that its way better then the old windows 10 version and EXTREMLY better then the buggy windows 11 alt tab, as it it shows built in minimized to tray uwp apps and doesn't really go by the rules of minimized to tray applications. 
It makes win11 usable without losing features that was on win10. 
 alt tab: 
 ![LmlQToW 1](https://user-images.githubusercontent.com/53096166/138576581-075f4c06-06b8-4833-9fb9-128065b6edd3.png)

Win tab: 
![hf1Cg98 1](https://user-images.githubusercontent.com/53096166/138576628-118dcc38-3e41-4168-abd0-278813e0d3f3.jpeg)

**Reply by valinet** [reply]
Yeah, it's been an issue, I have fixed it, I will push an update soon. Thanks.

**Comment by Poopooracoocoo** [comment]
Great work!!! Does this happen to have a scrollbar for when there are a lot of windows? ^.^

**Reply by valinet** [reply]
You continue tabbing and the rest of the windows start showing as well.

**Reply by Poopooracoocoo** [reply]
That's great. 
I was thinking more something for touch and mouse input, as well as for an easier experience when using a keyboard

**Reply by valinet** [reply]
Yeah, well, Alt-Tab is a keyboard driven interface primarily. Even if you have a touch screen, you can’t get there without pressing Alt+Tab. Everything is custom drawn, there’s no logic currently for drawing half rows etc. At some point, maybe that gets added in, but at the moment, I am focusing on improving the core experience and fixing some bugs.

**Reply by Poopooracoocoo** [reply]
Right on. I respect that. 
I get that alt-tab is keyboard driven, but combinations of input methods are especially possible on laptops.

**Reply by valinet** [reply]
I understand what you are saying either, but also, I have limited time to work on this. Also, I don't necessarily do a general public prioritization of features, but rather implement what I like and feel like implementing. The code is public and written okay, I'd say, it would more than welcome a contribution on this area just fine. Anyway, thanks for the tip, I will think about it.

**Comment by M0rfic** [comment]
Hi,

Would it be possible to make it display only the windows on the active monitor (the one with the mouse or where the switcher is activated), similar to how the taskbars show buttons for the windows corresponding to the monitor they're on?

And if I may, a second question off-topic though (but didn't really want to start a new thread for it): any chance to have the systray appear on secondary taskbars as well?

Kind regards

**Reply by valinet** [reply]
Hi

1. It's possible. To me, that kind of defeats the purpose of the window switcher (i.e. showing you all tasks in chronological order), but better than arguing why I personally don't find much use for that, it was actually quicker to implement it and end the story. So yeah, it's there, will eventually come.
2. Slim. The main and secondary taskbars are 2 different classes with 2 different implementations mainly. Explorer generates a single system tray and places it on the main taskbar, from the info I gathered while "succinctly" looking on the disassembly. Idk, there are some utilities like Actual Multiple Monitors that have reimplemented a more powerful taskbar, you could try those as well if this is an urgent must.

Thanks

**Reply by M0rfic** [reply]
Hi Valentin,

1. Thanks for taking the time to reply and also implement that CR, I really appreciate it. Just FYI (you probably don't care but I didn't give you any context in the first place), I have a multi monitor setup and several instances of my apps spread across them, and being able to quickly switch to a specific one is very useful to me. So thanks again.
2. I see, no it's nothing urgent so if there's any chance I'll gladly wait, even if it's slim (please do not take this as some sort of "reverse psychology demand", I'm not trying to be rude or anything). You're quite right, I've been using Actual Window Manager (includes AMM) for roughly 10 years now and it's been very helpful. However lately their release cycle has been getting slower and slower, to the point of ~1 release per year, and the current version is not "exactly" W11 compatible (not complaining, just saying), hence my question here.

Kind regards

**Reply by Gaurav-Original-ClassicShellTester** [reply]
@M0rfic Hi. I use AMM too. Have you noticed anything broken? I am sure they will update it as it's been around for so long and has remained the only non-bloated app for multiple monitors with ideal desired features working the way they should (or how Microsoft should have handled the multimon experience in Windows).

**Reply by M0rfic** [reply]
@Gaurav-Original-ClassicShellTester I've only recently switch to W11 so I didn't get the chance to give it a decent round of testing to gather comprehensive information. Nonetheless, so far on my work laptop AWM warns me that the Taskbar Manager fails to start repeatedly (even though I disabled, at least theoretically, the option in the GUI for now) so no taskbars. Could be some sort of config issue after upgrading to W11 but I have yet to investigate. On my PC until yesterday, the start button and the action center button were not working. After updating to the latest build last night (22000.282 if I'm not mistaking) they seem to work, not sure if it has anything to do with ExplorerPatcher which I applied yesterday or not (although the start button has some minor freezes once in a while). Systray does not have audio & network buttons. I think this pretty much sums it up so far.

I do hope they release some updates for W11, although I have yet to see an official announcement (found only [this issue](https://www.actualtools.com/forum/read.php?FID=10&TID=5724) so far mentioning they will start working on W11 support), and also the latest release for AWM was on 12 Feb 2021. Perhaps the release was flawless, who knows?!

Anyway, I'll see what I can crop up using all of these combinations of software, but at least for now I have a decent environment thanks to ExplorerPatcher, and I feel like I should buy this man a beer 😄 .

Cheers

**Comment by remyblok** [comment]
I love the windows 10 Switcher. It so much better then the windows 11 one. I do have one request though regarding the max width. At home I have an ultra width monitor (5120x1440) so I have set the max with to 55%. But when I unhook my laptop from the display and go back to a 1080p display 55% width is really weird. So I find myself switching the width.

I was wondering if you can add a second max width option in pixels as an absolute max width. So the maximum width as a percentage is calculated, But if this is larger then the absolute width in pixels it just takes that value. That way I can keep it at 80% and works well for normal displays. And set the absolute max width to 2560px, which then will keep the width in check on my ultra wide.

Thanks in advance should you choose to add this feature.
In any case keep up the great work!

**Reply by valinet** [reply]
I understand, interesting situation, a more general approach should be sought ideally, but as a quick fix, this is actually pretty good and okay. Implemented in the latest commit. A pre-release (.5) should be built and published soon. Please test with that. Configure `MaxWidthAbs` and/or `MaxHeightAbs` DWORD values in `HKCU\Software\Microsoft\Windows\CurrentVersion\Explorer\ExplorerPatcher\sws`. The formula is as you specified, thank you very much for posting exact guidance, it eases and shortens the implementation time a lot.

Thanks for the kind words!

**Reply by remyblok** [reply]
Thank you for the quick response and implementation! I've tested this a bit and works like a charm!

A more general approach is not that easy I guess. The most logic option would be to have a set of settings per monitor-setup or screen resolution. A bit like how FancyZones in PowerToys allows you to select a different layout per monitor. But this would make the settings window much more complex. 

This is pretty much an edge case. I don't know is there are other settings that would benefit from such a setup. So for now I think that this quick fix is sufficient.

**Reply by valinet** [reply]
Yeah, me too, I like it for now as well, really handy indeed. Thanks.

**Comment by BraINstinct0** [comment]
Pressing the X button does not close any window(not working): is this issue solvable, or probably happening on my side only? I wasn't sure about this even after reading the notes above, so leaving a question here.

**Reply by valinet** [reply]
It works, try not to move the mouse between mouse down and mouse up. The way it is coded now, the point has to be exact for it to work. Not ideal, I will look into it.

**Reply by valinet** [reply]
I improved on that in 39.1, should now be easily working as expected.

**Reply by BraINstinct0** [reply]
Can confirm it is working as intended(intuitively).👍👍👍👍👍

**Comment by kocane** [comment]
Beautiful work, really. Could you perhaps make it so that windows can be closed with mouse 3 button as in the vanilla window switcher?

Also, how would you feel about making the SWS a standalone modification?

**Reply by valinet** [reply]
I implemented middle button closing in the latest pre-release (40.9).

`sws` is actually a separate repository. The project can be built standalone, it creates a library which then has to be loaded by *some* executable and it *works*. The idea was for interested folks to take that and integrate it into their existing projects with ease. One such project that implements `sws` is, indeed, ExplorerPatcher.

There are a couple of issues `sws` does not currently build as a fully fledged executable:

* The burden of maintaining yet another application. The thing is, I use ExplorerPatcher as well, and that already takes a lot of time to develop, but since I already use it, integrating this in there was natural. I personally don't have yet more time to devote to testing `sws` in its standalone configuration. Logistically, as I said, it is doable, but if I cannot do the testing and fixing, then someone else has to do it. When I find such an interested person, then this will be one step closer to becoming a reality.
* There are a couple of implementation details that *I think* may prevent `sws` from fully working as a standalone executable at the moment; they are fixable, to some extent at least, but they are here at this moment, nevertheless:
  * To display in front of some applications, like Task Manager with always on option, On-Screen Keyboard etc, `sws` uses the restricted `CreateWindowInBand` API. This works only when used from Microsoft-signed executables. Now it works because `sws` currently runs in `explorer`, which of course is Microsoft-signed and has all kinds of other privileges and perks attached to it as well, which come in handy for the kind of work `sws` does. `explorer` is already "prepared" for hosting the window switcher, as it already does host the vanilla window switcher, as you called it. So, to have this thing work, `sws` would still have to run in a Microsoft-signed process. That's not an issue, I have done that in the past, there are mainly 2 other ways to do that:
    * Launch a `RuntimeBroker.exe` process, `CreateRemoteThread` in it with `sws` to run in that. Prone to `LoadLibraryW` failing due to mysterious reasons, unless shell code that bypasses that need is developed. Pretty robust solution.
    * Exploit the library search order of `ApplicationFrameHost.exe`. Same as with ExplorerPatcher, name `sws`'s library `dxgi.dll` and put it beside a [specific version of `ApplicationFrameHost.exe`](https://github.com/ModernFlyouts-Community/ModernFlyouts/raw/main/ModernFlyoutsHost/ModernFlyoutsHost_x64.exe). Not the latest one, because that does not load `dxgi.dll` anymore, but version 10.0.17763.1, similar to what [ModernFlyouts](https://github.com/ModernFlyouts-Community/ModernFlyouts) does (that's how their flyouts shows on the lock screen, they wouldn't be able to do it without using these restricted APIs). I favor this variant as it is much more reliable, but of course, there's the issue of redistributing that Microsoft binary.
  * I don't know if the calls to certain APIs would still be allowed, or be able to bypass [User Interface Privilege Isolation](https://en.wikipedia.org/wiki/User_Interface_Privilege_Isolation) from a non-`explorer.exe` process. I haven't tested. Things that come to my mind:
    * To switch to the selected window, `sws` uses the `SwitchToThisWindow` API call with the second parameter to `TRUE`. That indicates to the OS that the switch was initiated from the window switcher. This API is public, available for any application to call, but I don't think the OS is that dumb as to not check what the source of the call is and whether it indeed can be a window switcher or not.
    * The reliable way to close a window is `PostMessageW(hWnd, WM_SYSCOMMAND, SC_CLOSE, 0);`. When called from `explorer.exe`, this seems to reach elevated applications as well, which is what we want, but, again, idk, if it works from any regular process.
    * Notifications for language switching are received by using the reverse engineered APIs from `InputSwitch.dll`. Again, not tested outside of `explorer.exe`.

So yeah, there is a fair bit of testing to be done, a best strategy for how this should run has to be chosen (for example, you could still have it injected in `explorer`, but then, if still using `dxgi.dll`, then it means it's either `sws` or ExplorerPatcher, or the 2 have to "learn" to load each other), then people will demand an installer, automatic updates, then there's the mater of where you direct issues... at the moment, `sws` is discused in ExplorerPatcher, which I prefer since everything's in one place, but the correct way to do it would be to have `sws`-related discussions in that repository on GitHub.

There is a fair bit of maintenance and work to do for supporting this in yet again another configuration, and like, GitHub is a site for developers. The expectation is interested folks come, clone the source and use it however they want, modify it for their needs etc. I cannot, as a single person, support so many use cases and configurations etc. I don't really enjoy it either. User-facing functionality, like the configuration GUI, is not something that I fancy developing. Most of the work for having `sws` run "standalone" as you called it would be put into developing and maintaining yet another set of user-facing functionality, not actually improving on the core experience, and I largely see that as a big waste of time, for me personally. If someone else wants to do it, I mean, I don't even have to sanction them, I have nothing against that, they can just fork the project and simply do it. The infrastructure for that is available today, and I am open to that, no problem.

Also, many people seem to miss this, but the great thing about having access to source code is that you can configure absolutely everything to your liking, and with ease if the software is particularly written in that direction. Take `sws` for example, only a subset of parameters are exposed via the Properties GUI in ExplorerPatcher. Many more are specified in the [`sws_def.h`](https://github.com/valinet/sws/blob/master/SimpleWindowSwitcher/sws_def.h) file. The easiest way to customize those, easier than with any UI in this world which would need to constantly be updated to keep pace with that, is to just change the values in the file and recompile the software. It doesn't get easier or more efficient than that. And compiling is very easy with projects nowadays, you just install Visual Studio, `git clone` the project, double click the solution file, hit F5 and that's it.

**Reply by plofhaan** [reply]
I just tried 40.9, it doesn't work for me (fails to start explorer). I can manually start explorer by going in task manager, but not all my tray icons are loaded and TranslucentTB (UWP) doesn't start. Next, if I go in EP settings to restart explorer, it fails to start explorer again. The screen stays blank while only EP settings stays visible and each time when I click to restart explorer again, ~~I can see explorer starting and crashing immediately. I have verified that explorer crashes~~, by going in perfmon /rel. I reverted back to 40.8 for now.
**EDIT:** this issue is fixed in 22000.376.40.15.

**Reply by valinet** [reply]
@plofhaan Can you try with this?
[ep_setup.zip](https://github.com/valinet/ExplorerPatcher/files/7698971/ep_setup.zip)

**Reply by valinet** [reply]
@plofhaan Is that working ???

**Reply by plofhaan** [reply]
At first it didn't work and had to uninstall it. It asked me to start explorer next, so I did that. I reinstalled it and got the blank screen again. I manually started explorer again after that, and it appeared to be working. I also verified that I could use EP settings to restart explorer. Finally, I went in EP settings to update EP. Now it appears to be working and I have verified that I can still use EP settings to restart explorer.

**Reply by valinet** [reply]
So you're on 40.9 official now, right?

I don't think it is a problem with 40.9 specifically. That introduces almost nothing compared to 40.8. The zip I sent you is 40.8 with just the thing in `sws` as requested by @kocane. So basically 40.8. 40.9 is that plus I made the `sws` thread wait until the taskbar is created before enabling the switcher. I refuse to believe that small thing dynamited your whole setup. It doesn't make any sense to me, I know about lateral effects, but still... I think maybe you just did not experience this with 40.8 and it just happened on 40.9, and the actual changes between the 2 are irrelevant.

> It asked me to start explorer next, so I did that.

It does that when it services the application and Explorer was not running when it started doing that. It's there so that Explorer does not get launched elevated, which *may* be a security risk.

**Reply by plofhaan** [reply]
Yes, I'm on the official 40.9 pre-release now, as the updating went fine. The reason why I had to uninstall EP after I had installed the one from the zipfile that you posted was because typing in explorer.exe in task manager after I got to the blank screen did nothing, and, if it doesn't make sense to you, then it makes even much less sense to me. :]

**Reply by kocane** [reply]
@valinet Thanks for the elaborative reply! The extra work that comes with having to maintain two separate applications is an understandable reason for not wanting to make a seperate SWS application.  Also, thanks for the update - closing with M3 works like a charm.

Does using Startallback together with ExplorerPatcher cause any issues, to your knowledge? Because, I mean, if thats not the case then I have no problem with coupling the two together. What I did was to just select "Windows 11" taskbar in ExplorerPatcher and then SAB seemed to take over. 

The only thing I've noticed is that hovering over the close button on a thumbnail windows looks a bit odd. This doesn't appear without EP, but since I'm probably just asking for issues when combining the two programs, I don't exactly want to file an issue about it.
![image](https://user-images.githubusercontent.com/7694813/145718994-20c2d09b-8178-4ebd-913b-59de088c2a38.png)

It's not ideal but I kind of _need_ Simple Window Switcher, so it is what it is. Some of the other functionality EP adds, is nice too ;)

**Reply by plofhaan** [reply]
@kocane I haven't tried SAB, I only use EP with Open-Shell, Battery Mode, TranslucentTB and 7+ Taskbar Tweaker. The only thing I can think of that is still missing in EP is:
`reg add HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Explorer\Advanced /f /v ShowSecondsInSystemClock /t REG_DWORD /d 1`
...but there's no more room left for @valinet to add it to the Taskbar section of EP settings. ;)

**Reply by valinet** [reply]
@plofhaan Yeah, honestly, I don't really know what may have caused that, but I am pretty sure it just did not occur with 40.8 and older, it wasn't introduced by the latest update. Btw:

![image](https://user-images.githubusercontent.com/6503598/145719726-0f8f7868-5760-4c17-8a6c-ff14590053aa.png)

@kocane You're welcome.

As I mentioned above, I don't test with all the configurations in the world. I don't use StartAllBack, I don't routinely test against and alongside that. Yeah, both programs kind of overlap in the hooks they set, but indeed, when setting "Windows 11" taskbar in EP, *most* of its hooks are not set anymore. Also, EP starts first, before SAB, so all it sets should be safely overriden by SAB if it needs to. That being said, I heard about this thing you showed, it was mentioned by someone else in some other thread before. I did not look into it, like, if you were me, would you? There are 2 strategies to figuring that out, each more "enjoyable" than the other:

* Compile EP from source. Selectively take out stuff it sets up, hooks etc until the visual glitch stops. You then narrow it down to the exact hook that causes this and spend from minutes to days depending on your luck and inspiration trying to find a workaround so that it does not happen anymore.
* Compile sws into a standalone application as I described above, fixing all incompatibilities and use that. Again, depending on luck, inspiration and decisions taken, takes from hours to weeks (???)

There are options, there are tools, all that is (still) required is a human to put them all together. You can try option 1 above, maybe you figure it out. Then, armed with at least the knowledge of what exactly is making that happen, we can write a patch to try to mitigate it, and pray it doesn't break someone's workflow in some mysterious but totally possible way. Filing an issue doesn't help much, actually solving it would be even more helpful. This is more easily done here, as everyone has access to the source code, as opposed to SAB, where contacting the developer is the only realistic option for most things.

**Reply by plofhaan** [reply]
> ![image](https://user-images.githubusercontent.com/6503598/145719726-0f8f7868-5760-4c17-8a6c-ff14590053aa.png)

Damn. I must be going blind... 😆

**Reply by kocane** [reply]
@valinet  I wish I was worth anything when it came to _actually_ contributing with code, but I'm worthless there so all I can do it's complain and point out issues. There's clearly just issues with combining the two and I understand if it's not exactly something of high priority to be fixed. The features that EP brings outweights the issues that combining it with SAB brings by far, so... 

By the way, what would you think about adding the oldschool aero blur to the SWS? If it's something that still can be utilized. 

@sisericm it could be the theme I'm using with SAB - not sure! It don't bother me too much.

**Reply by valinet** [reply]
What “oldschool Aero blur”? Again, a screenshot doesn’t hurt. Aero Glass has been disabled since Windows 8 and kind of removed for good from dwm since Windows 10 2004. Anyway, officially, “DwmExtendFrameIntoClientArea” just extends the solid color of the title bar into the client area since Windows 8, and other than that, “DwmEnableBlurBehind” is also non-functional, so idk what else is still in there.

> There's clearly just issues with combining the two 

I disagree with that.

**Reply by kocane** [reply]
I meant the one from Windows Vista and 7. Yup, sounds like it has been removed then, nevermind it. 

>I disagree with that.

Why? It might be minor issues (and they may not very clear, actually) like that close button on thumbnails in the taskbar, that I brought up earlier, or the missing transaprency in the context menu like @realityisgay, for some reason, brought up in this thread. Not that I mind it.

**Reply by valinet** [reply]
> Why?

Because it’s not quote “just issues” end quote as you said, but there are some “issues” as you all call them, but they generally work just fine side by side.

**Reply by kocane** [reply]
Poor wording, I agree. I also agree that they work fine side by side :-)

**Comment by sisericm** [comment]
Hi @kocane,
This is what I get using EP and Openshell but no SAB. Admitttedly, I'm using EP Win10 taskbar to provide the preview functionality. 
It's just a guess but could the strange button alignment be the result of SAB drawing (or partly drawing) its own previews rather than utilising the windows previews? 

![Image 001](https://user-images.githubusercontent.com/8255994/145766745-69126b4c-877c-42d8-b9df-44319f2f060c.png)

**Comment by PinchToDebug** [comment]
Can i use this standalone somehow?

**Reply by valinet** [reply]
https://github.com/valinet/ExplorerPatcher/discussions/110#discussioncomment-1793568

**Reply by PinchToDebug** [reply]
I'm using startallback at the same time and it ExplorerPatcher makes my context menu opaque, is there a way to disable this feature from ExplorerPatcher?

**Reply by valinet** [reply]
I don’t understand why recently everyone posts everything in this thread. 

Anyway, idk what you mean. Spend a bit more than 5 minutes with the program and maybe you find the answer yourself even. Like, I don’t understand, you mean the Explorer context menu in folders? Can’t you leave it untouched (so the crappy Windows 11 new menu shows, and also not disable immersive menus) from ExplorerPatcher and then choose your options in StartAllBack? Doesn’t that work? What have you tried? 

And I don’t understand why it’s always ExplorerPatcher that does something bad, it’s always ExplorerPatcher that does something negative or gets into someone’s way, never the million other third party software they are using, it’s always ExplorerPatcher that has to workaround because of something in Windows or even in 3rd party software… that never can be changed.

**Reply by PinchToDebug** [reply]
I tried everything otherwise I wouldn't have ask for help. I've been tinkering w it for at least 2 hours.
I tried installing the latest pre-release and the stable version too disabling / enabling almost every related settings and then restarting my explorer many many times.

**Reply by PinchToDebug** [reply]
I'm talking about the context menus on the desktop and in the File Explorer
![StartAllBack](https://user-images.githubusercontent.com/52333417/145935544-39d96109-bd6d-4717-9693-9e27984f9e17.png)
![sad](https://user-images.githubusercontent.com/52333417/145935551-85f51256-57e6-44c1-8fd9-d46df8e0c817.png)

**Reply by PinchToDebug** [reply]
It's loosing all the transparency no matter what I do with the settings.

**Reply by valinet** [reply]
Then idk what else to suggest, I don’t test every configuration. Try in a clean VM, install EP, then SAB, and enable classic context menus from SAB and see what you get.

**Comment by Amit-Shan** [comment]
> * Microsoft Edge tabs as windows, because I HATE that feature. `Alt`+`Tab` is for windows only.

Well, I certainly hope you change your mind. In my opinion, "Win+Tab" is for windows only. At least, that is what makes sense to me and is a matter of preference anyway, probably why Windows has an option to disable browser tabs in Alt+Tab. I think that is the default behavior as well. But this is a free software, so it is entirely up to you. 

I hope you make a paid version of the software someday with requested features you might not like personally. I would buy it in a heartbeat to be honest..

**Comment by Amit-Shan** [comment]
> * Microsoft Edge tabs as windows, because I HATE that feature. `Alt`+`Tab` is for windows only.

Well, I certainly hope you change your mind. In my opinion, "Win+Tab" is for windows only. At least, that is what makes sense to me and is a matter of preference anyway, probably why Windows has an option to disable browser tabs in Alt+Tab.

**Reply by valinet** [reply]
Unlikely. Also, there isn’t any API for that. And in no way I am also reverse engineering and implementing this from a political point of view unless they make a public API 3rd party apps can subscribe to and use in a similar fashion. I am tired of alternatives being discriminated for no reason.

> In my opinion, "Win+Tab" is for windows only.

What is Alt-Tab for then? Random things that presumably any app can throw in there? Alt-Tab has traditionally been a fast list of top-level windows. Just that. Win-Tab is a much newer attempt at implementing an Expose like feature, where you see, again, all the top-level windows at a glance. What I think would work best there is an implementation of GNOME Activities; I plan to implement that soon, but it obviously will require a bit more effort to do right. I am using sws as a learning tool towards implementing the bigger stuff. Oh, and also, it takes ages, time flies too fast.

**Comment by anderspk** [comment]
Hello,
Just came over this project, and I think it looks really great!
My main issue with alt tab at the moment, is that I would love to be able to group results by instance. Currently, when using alt + tab with several open instances of applications like vscode and chrome, it can become challenging to quickly identify and switch to a specific window - especially when the alt tab interface is populated with a large number of windows.

What I would love, is a grouping functionality within the alt tab interface. This would involve grouping windows by their application type. For example, all vscode windows would be grouped together, as would all Chrome windows. This could be visually represented as a grid, with each row corresponding to an application type and columns representing individual instances of that application. Users could navigate between application types using the left/right arrow keys and select a specific instance using the up/down keys.

Is this a feature that sounds interesting to you?

As a side note, if you _do_ find it interesting, I also think it would be an amazing addition to Windows Powertoys 😁

**Reply by pyrates999** [reply]
Unless an existing API allows to do this, then I would say no.

**Comment by epikgit** [comment]
Hello Valentin,
Thanks for your terrific program. Without your effort I would not have installed Win11 for sure.

Regarding the Windows switcher I have an idea for a little improvement.
Why not add (optionally) a little number on any window thumbnail according to the monitor in which it resides?
I use three monitors and every time I switch the current window I don't know where the active window has just appeared (poor lateral sight) so I have to turn my head on the left and on the right until I find the new window. 
Plus if this window is similar to another already opened I have to squeeze my neurons to understand which one is the last opened.

If I have the little number on any thumbnail I see on the windows switcher, I'll know in advance where the window pops up and I can turn my head accordingly.

**Comment by ne0YT** [comment]
@valinet 

I want to use this as a standalone.exe is it smh possible? I don't run explorer.exe
